gui/mozregui/log_report.py

```diff
@@ -27,15 +46,44 @@ def on_log_received(self, data):
         cursor_to_add = QTextCursor(message_document)
         cursor_to_add.movePosition(cursor_to_add.End)
         cursor_to_add.insertText(log_message + '\n')
+
         if data['level'] in COLORS:
             fmt = QTextCharFormat()
             fmt.setForeground(COLORS[data['level']])
             cursor_to_add.movePosition(cursor_to_add.PreviousBlock)
+            log_lvl_data = LogLevelData(log_levels[data['level'].upper()])
+            cursor_to_add.block().setUserData(log_lvl_data)
             cursor_to_add_fmt = message_document.find(data['level'],
                                                       cursor_to_add.position())
             cursor_to_add_fmt.mergeCharFormat(fmt)
         self.ensureCursorVisible()
 
+    @Slot()
+    def on_custom_context_menu_requested(self):
+        menu = QMenu(self)
+        for action in self.actions:
+            menu.addAction(action)
+        #self.contextMenu.move(self.cursor().pos())
+        menu.popup(self.cursor().pos())
+
+    @Slot()
+    def on_log_filter(self):
+        log_lvl_name = str(self.sender().iconText()).upper()
+        log_lvl = log_levels[log_lvl_name]
+        cursor = QTextCursor(self.document())
+        current_block = cursor.block()
+        while True:
```

Comment:
Hm, I don't like this while True with the break, since it is based on the presence or not of the user data. You should be able to iterate over all the block with a syntax like:
```python
while current_block.isValid():
    # do stuff
    current_block = current_block.next()
```
This way you are sure that you get every block.